src/timing.c

```diff
@@ -113,18 +135,124 @@ void jl_destroy_timing(void)
     jl_ptls_t ptls = jl_current_task->ptls;
     jl_timing_block_t *stack = ptls->timing_stack;
     while (stack) {
-        _jl_timing_block_destroy(stack);
+        jl_timing_block_end(stack);
         stack = stack->prev;
     }
 }
 
-jl_timing_block_t *jl_pop_timing_block(jl_timing_block_t *cur_block)
+static const int get_timing_subsystem(const char *subsystem) {
+    const char **match = (const char **)bsearch(
+        subsystem, jl_timing_subsystems, JL_TIMING_SUBSYSTEM_LAST,
+        sizeof(const char *), (int(*)(const void*, const void*))strcmp
+    );
+    if (!match)
+        return JL_TIMING_SUBSYSTEM_LAST;
+
+    return (int)(match - &jl_timing_subsystems[0]);
+}
+
+// This function is analogous to __itt_event_create but for the counts backend
+//
+// `event` is required to live forever
+JL_DLLEXPORT jl_timing_counts_event_t *jl_timing_counts_event_create(const char *event) {
+#ifdef USE_TIMING_COUNTS
+    const size_t n = jl_timing_counts_events.len;
+    for (size_t i = 0; i < n; i++) {
+        jl_timing_counts_event_t *other_event = (jl_timing_counts_event_t *)jl_timing_counts_events.items[i];
+        if (strcmp(event, other_event->name) == 0)
+            return other_event;
+    }
+
+    // No matching event found - create a new one
+    jl_timing_counts_event_t *new_event = (jl_timing_counts_event_t *)malloc(sizeof(jl_timing_counts_event_t));
+    arraylist_push(&jl_timing_counts_events, (void *)new_event);
+    new_event->name = event;
+    jl_atomic_store_relaxed(&new_event->total, 0);
+    return new_event;
+#else
+    return NULL;
+#endif
+}
+
+JL_DLLEXPORT jl_timing_event_t *jl_timing_event_create(const char *subsystem, const char *name, const char *function, const char *file, int line, int color) {
```

Comment:
I feel like we should deduplicate at this point, rather than relying on each individual event system to handle deduplication. I think at this point we should only have a 1:1 mapping of event name to source location, so there isn't a reason to only deduplicate for counts.

Reply:
We already de-duplicate at the source location level, but that's the caller's responsibility

For any de-duplication that goes beyond that, I think splitting to the back-ends makes sense since they all use slightly different information.

For counts specifically, being able to re-use event names in multiple locations seems like a nice feature, and the de-duplication gives us the flexibility to group counts by subsystem in the future if we want to, so I don't really want to rip out the counts event de-dupe either.